### PR TITLE
Add react context for user and session

### LINKED ISSUE #15: Add front-end JWT session token persistence
Authentication with the server is done using a JWT token. The front-end needs to store this between components etc. Suggestion to use [context](https://reactjs.org/docs/context.html).

### PR DESCRIPTION
**Pull request by joshxg**
fixes #15 

Added a context for User and Session. Chose two separate contexts to provide different components with only the data they require. Also chose to use HOC's to wrap components and provide the data as it seems cleaner than wrapping each component and cluttering the JSX.

Not sure if this approach is the best or if the placement of files within the file structure is best?

### CODE REVIEW

**Comment by joshxg** [comment]
Has been updated with development - loginmodal will be updated with issue #14

File: src/ui/MSupplyRnR.js [hunk 1]
```diff
@@ -1,40 +1,80 @@
...
-    this.setState({ loginModalOpen: false });
-    // set withApp state to be session id.
+    this.setState({ loginModalOpen: false, sessionId, user });
   };
 
   render() {
-    const { loginModalOpen } = this.state;
+    const { classes } = this.props;
+    const { loginModalOpen, user, sessionId } = this.state;
     return (
```
Comment (wlthomson) [inline src/ui/MSupplyRnR.js]: Does this mix style with logic? Suggest explicit control statement based on login state, either in render or by settings the components to be returned as state, simplest way would be something like `if (loggedIn) { return (...) } else { return (...) }`.

File: src/ui/MSupplyRnR.js [hunk 2]
```diff
@@ -1,40 +1,80 @@
...
+
+          <SessionContext.Provider value={sessionId}>
+            <div className={classes.content}>
+              <RequisitionsPage
+                drawerOpenHandler={this.drawerHandler}
+                createHandler={this.createRequisition}
+              />
+            </div>
+          </SessionContext.Provider>
+        </div>
```
Comment (wlthomson) [inline src/ui/MSupplyRnR.js]: Does this mix style with logic? Suggest explicit control statement based on login state, either in render or by maintaining the child components to be returned as parent component state, simplest way would be something like:

```
if (loggedIn) 
  return (
    <div> 
    // UserContext etc.
    </div>
  )
return <LoginModal.../>
```

File: src/ui/MSupplyRnR.js [hunk 3]
```diff
@@ -1,40 +1,80 @@
...
         <LoginModal
           open={loginModalOpen}
-          closeHandler={this.handleModalClose}
+          closeHandler={this.handleSuccesfulLogin}
         />
```
Comment (wlthomson) [inline src/ui/MSupplyRnR.js]: Does this mix style with logic? Suggest explicit control statement based on login state, either in render or by maintaining the child components to be returned as parent component state, simplest way would be something like:

```
if (loggedIn) 
  return (
    <div> 
    // UserContext etc.
    </div>
  )
return <LoginModal.../>
```

File: src/ui/context/SessionContext.js [hunk 4]
```diff
@@ -0,0 +1,5 @@
+import React from "react";
+
+const SessionContext = React.createContext();
+
+export default SessionContext;
```
Comment (wlthomson) [inline src/ui/context/SessionContext.js]: Is it necessary to have a separate file for `SessionContext` and `UserContext`? Could these be included with their respective HOCs?

**Comment by wlthomson** [comment]
Looks good. Understand reasoning for different contexts, but as session and user are tightly bound and 1-to-1, it might be unnecessary. JWT returned by the server can also include user, in that way only one context is needed with the JWT token. This has the added benefit of being immutable, so it can be passed around components without worrying about side-effects etc.

File: src/ui/context/SessionContext.js (hunk 4 above)
Comment (joshxg) [reply src/ui/context/SessionContext.js]: I think combining the context's could be good, but I don't like combining everything. Potentially the HOC's could be in their own folder - ui/hoc as there is potential for more HOC's in the future?

Comment (wlthomson) [reply src/ui/context/SessionContext.js]: As `ErrorBoundary` is in its own folder, this sounds appropriate. Maybe keep the contexts in `ui/context`, but the respective HOCs under `ui/components/hoc`.

**Comment by joshxg** [comment]
@wlthomson I like the idea of components only having access to the data they need. Also, without a setter, the context is immutable. 

The username can be retrieved easily from the users input, rather than being within the JWT/body of a response from /login.

I'm not married to it, but my thoughts for the future made me plan it like this: a pure auth context can be passed to components which make requests. The usercontext could potentially be expanded upon in the future as maybe a 'navbarContext'/info context which could potentially contain additional information - eg notifications, last logged in date, messages etc (maybe bad examples but hopefully you get the point)

**Comment by wlthomson** [comment]
@joshxg I see the argument, my only concern is that with two contexts we are duplicating state and having multiple sources of truth. It would be good if we can keep the JWT as the sole source of truth as much as possible, as the ui should be as "dumb" as possible. For example, last logged in date etc. will all have to come from the server, and the server will communicate all its data using JWT.

**Comment by wlthomson** [comment]
@joshxg After discussion, agree that JWT can be decoded and used to set app context using multiple context providers. Solution is to ensure state between these providers is kept synced. As JWT does not update after login, perhaps a library such as immutable.js may be suitable.

**Pull request merged**